rfcs/text/0000-dss-deletion-daemon.md

```diff
@@ -0,0 +1,268 @@
+   The bundle UUID and version is required for deletion. The admin CLI will return a list of files that will be 
+   deleted and bundles that are effected by this deletion. Bundles may be affected if they share files or are derived 
+   from the deleted bundle.
+   
+   ```bash
+   delete_bundle.py <uuid>.<version> --reason [consent_withdrawn, consent_absent, service_disruption, legal] 
+    --type [logical, physical] --details "<additional info>"
+   
+   Files to be deleted:
+   <fileblobs>
+   .
+   .
+   .
+
+   Effected Bundles:
+   <uuid>.<version>
+   .
+   .
+   .
+
+   confirm (yes or no)?
+
+   ```
+
+3) **Logical Deletion:** The DSS places a **bundle tombstone** in the
+   underlying storage (S3 or GCP bucket). The tombstone's content is as follows:
+   ```JSON
+   { "reason": "<reason>",
+     "details": "<additional info>",
+     "admin_email": "<admin>"
+   }
+   ```
+   The tombstone causes `GET /bundle/{uuid}` to return a 404. If `{version}` 
+   signifies the latest version of the bundle
+   identified by `{uuid}`, requests for `GET /bundle/{uuid}` without a version
+   also return a 404. The same applies to HEAD requests, both versioned and
+   unversioned. The file versions referenced by the deleted bundle version are
+   not immediately affected by the deletion and remain accessible. The bundle and associated files are added to 
+   a deletion queue for regularly schedule **Physical Deletion**.
+   
+   A notification is sent out to subscribers of datastore deletions when a new bundle tombstone is created, so they can 
+   update their index appropriately.
+
+4) On a regular schedule the deletion queue is processed by code running within a secure boundary. An administrator 
+   may manually invoke the deletion process or wait for it run at it's scheduled time once daily.
+   The deletion daemon is responsible for performing **Physical Deletes** of data from the deletion queue. 
+   A physical delete makes all files associated with a bundle or bundle version inaccessible. This will make HEAD and 
+   GET requests against those file versions return a 404.
+   
+   A log entry is produced when ever the deletion daemon is run. It contains information on what bundles and files
+   have been logically deleted, the reason, the admin who executed the deletion, and info about the deletion markers. 
+   The same information is stored in the **Deletion Table**. The deletion table tracks all bundles and files currently
+   pending permanent deletion. The Deletion Table contains enough information to reverse Physical Deletes before 
+   the data is Permanently deleted.
+
+5) Data that has been physical deleted remain in the bucket for 7 days before being permanently deleted from the bucket.
+   This grace period allows the deletion to be reversed within that time period. After the grace period the data will 
+   be permanently deleted.
```

Comment:
I don't believe a deletion bucket is necessary. Instead, data that is scheduled for physical deletion should be made inaccessible (logical deletion) and remain in the main bucket for a grace period.

Introducing a deletion bucket forces us to unarchive and copy data in order to delete it. That process itself can be long and cumbersome, and I don't believe the benefits of having a deletion bucket outweigh the problems with it.

Reply:
this is an obfuscation of the underlying bucket lifecycle policies that will be used to clean up files that have been deleted in a version bucket.